apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/RdaLoadJob.java

```diff
@@ -0,0 +1,178 @@
+package gov.cms.bfd.pipeline.rda.grpc;
+
+import static gov.cms.bfd.pipeline.sharedutils.PipelineJobOutcome.NOTHING_TO_DO;
+
+import com.codahale.metrics.Meter;
+import com.codahale.metrics.MetricRegistry;
+import com.google.common.base.Preconditions;
+import gov.cms.bfd.pipeline.sharedutils.NullPipelineJobArguments;
+import gov.cms.bfd.pipeline.sharedutils.PipelineJob;
+import gov.cms.bfd.pipeline.sharedutils.PipelineJobOutcome;
+import gov.cms.bfd.pipeline.sharedutils.PipelineJobSchedule;
+import java.io.Serializable;
+import java.time.Duration;
+import java.time.temporal.ChronoUnit;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.concurrent.Callable;
+import java.util.concurrent.Semaphore;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * General PipelineJob instance that delegates the actual ETL work to two other objects. The
+ * RdaSource object handles communication with the source of incoming data. The RdaSink object
+ * handles communication with the ultimate storage system. The purpose of this class is to handle
+ * general PipelineJob semantics that are common to any source or sink.
+ *
+ * <p>Since the streaming service can run for extended periods of time this class is designed to be
+ * reentrant. If multiple threads invoke the call() method at the same time only the first thread
+ * will do any work. The other threads will all immediately return with an indication that they have
+ * no work to do.
+ */
+public final class RdaLoadJob<TResponse> implements PipelineJob<NullPipelineJobArguments> {
+  private static final Logger LOGGER = LoggerFactory.getLogger(RdaLoadJob.class);
+  public static final String CALLS_METER_NAME =
+      MetricRegistry.name(RdaLoadJob.class.getSimpleName(), "calls");
+  public static final String FAILURES_METER_NAME =
```

Comment:
> * where is the proto file loaded by the GrpcRdaSource?

In java the proto file is used at compile time to generate the stubs.  It's not needed at runtime.  A maven plugin finds all of the proto files and generates stubs during `compile` and `compile-custom` goals.  The stubs are written to the `target/generated-sources` directory.

> * its premature to implement any "replay" like mechanisms as we don't have a spec yet from the RDA team, but have you given any thought as to how that _might_ be accomplished going forward?  Some environments, specifically prod, I'm guessing will need to reference data from the sink in order to know at what point in time we should start streaming from

My thinking is that the replay token (whatever it turns out to be in the end from RDA, likely a 64-bit number) will be stored in the database.  Very likely it will be stored in a new column with each claim so we know what RDA API record generated the last database update to each claim record in the database.

Then I can see two likely ways to approach the restart.  I haven't given these a ton of thought but they both seem to fit pretty well:

- Have some sort of progress tracker that is independent of both the sink and source and is used by both to track progress and figure out where we left off in the stream.  This is nice in that progress is a cross cutting concern between the source and sink.
- Add a method to the `RdaSink` interface directly to retrieve an appropriate replay token to resume.  This has the nice property that the progress is easily written to the database in the same transaction as the claims themselves.

I don't feel strongly about either option at this point and there may be another, better one once we know more.


> the failures and successes meters are measuring the success/failure of the number of "calls" made, and not the messages processed. Did I interpret that correctly?

Yes at this level it's all about calls.  More detailed counts will be tracked in the lower level classes like `GrpcRdaSource` and whichever `RdaSink` implementation is being called.